lib/plugins/package/lib/packageService.js

```diff
@@ -17,45 +16,98 @@ module.exports = {
+  },
+
   getIncludes(include) {
     const packageIncludes = this.serverless.service.package.include || [];
     return _.union(packageIncludes, include);
   },
 
-  getRuntime(runtime) {
-    const defaultRuntime = 'nodejs12.x';
-    return runtime || this.serverless.service.provider.runtime || defaultRuntime;
+  async getExcludes(exclude, excludeDevDependencies, includeNonDefaultExcludes) {
+    const packageExcludes = this.serverless.service.package.exclude || [];
+    // add local service plugins Path
+    const pluginsLocalPath = this.serverless.pluginManager.parsePluginsObject(
+      this.serverless.service.plugins
+    ).localPath;
+    const localPathExcludes = pluginsLocalPath ? [pluginsLocalPath] : [];
+    // add layer paths
+    const layerExcludes = this.serverless.service
+      .getAllLayers()
+      .map(layer => `${this.serverless.service.getLayer(layer).path}/**`);
+    // add defaults for exclude
+    const configFilePath = await serverlessConfigFileUtils.getServerlessConfigFilePath(
+      this.serverless
+    );
+    const serverlessConfigFileExclude = configFilePath ? [path.basename(configFilePath)] : [];
+
+    const devDependencies = [true, undefined, null].includes(excludeDevDependencies)
+      ? await this.excludeDevDependencies()
+      : [];
+
+    if (includeNonDefaultExcludes) {
+      // Using old include / exclude. To be removed in version 3
+      return _.union(
+        this.defaultExcludes,
+        serverlessConfigFileExclude,
+        localPathExcludes,
+        packageExcludes,
+        layerExcludes,
+        devDependencies,
+        exclude
+      );
+    }
+
+    return _.union(
+      this.defaultExcludes,
+      serverlessConfigFileExclude,
+      localPathExcludes,
+      layerExcludes,
+      devDependencies
+    ).map(pattern => `!${pattern}`);
   },
 
-  getExcludes(exclude, excludeLayers) {
-    return serverlessConfigFileUtils
-      .getServerlessConfigFilePath(this.serverless)
-      .then(configFilePath => {
-        const packageExcludes = this.serverless.service.package.exclude || [];
-        // add local service plugins Path
-        const pluginsLocalPath = this.serverless.pluginManager.parsePluginsObject(
-          this.serverless.service.plugins
-        ).localPath;
-        const localPathExcludes = pluginsLocalPath ? [pluginsLocalPath] : [];
-        // add layer paths
-        const layerExcludes = excludeLayers
-          ? this.serverless.service
-              .getAllLayers()
-              .map(layer => `${this.serverless.service.getLayer(layer).path}/**`)
-          : [];
-        // add defaults for exclude
-
-        const serverlessConfigFileExclude = configFilePath ? [path.basename(configFilePath)] : [];
-
-        return _.union(
-          this.defaultExcludes,
-          serverlessConfigFileExclude,
-          localPathExcludes,
-          packageExcludes,
-          layerExcludes,
-          exclude
+  async getPatterns(packageConfig = {}, prefix = '') {
```

Comment:
That's essentially what I've done.

There are 3 functions that gather the patterns: `resolveFilePathsAll`, `resolveFilePathsFunction and `resolveFilePathsLayer`.

The three of them do the same:
```js
    return this.getExcludes(funcPackageConfig.exclude, true)
      .then(exclude => {
        const params = { exclude, include: this.getIncludes(funcPackageConfig.include) };
        return params;
      })
      .then(params => this.excludeDevDependencies(params))
      .then(params => this.resolveFilePathsFromPatterns(params));
```

So I've abstracted that logic to a `getPatterns` function which calls `getExcludes` and `getIncludes`.
`getIncludes` does exactly as you suggested.
`getExcludes` only change is that it adds the dependencies from `devDependencies`, instead of doing a separate call to the function in `zipService` that would add it to the `exclude` property as I explained in a different comment.

Does that make sense?
I understand that you want to minimize changes, but I think that the code was getting a little bit spaghetti and the implementation in this PR really simplifies things.

Reply:
> That's essentially what I've done.

Changes I see now, look as way more than needed to accomplish what's described above, and here we deal with really sensitive functionality, that's why I'm pushing for minimal needed changeset 

I wouldn't touch `getExcludes` and I don't think any changes are needed in `zipService`. I've elaborated on that a bit more here: https://github.com/serverless/serverless/pull/8581#discussion_r539490987

Also, sorry I don't want to sound dismissive, in overall you've done a really great work.